src/metaSources.js

```diff
@@ -0,0 +1,86 @@
+const sources = require("./sources");
+const utils = require("./utils");
+
+const PRODUCTION = process.env.NODE_ENV === "production";
+
+/**
+ * A meta-source representing all known groups.
+ * Created using a IIFE to close around state representing the
+ * available datasets & narratives for all groups. This allows
+ * us to update this state on our own timeframe, independent
+ * of API requests.
+ */
+const Groups = (() => {
+
+  const groupSources = new Map( // instances of each group Source
+    Array.from((sources))
+      .filter(([name, Source]) => Source.isGroup())  // eslint-disable-line no-unused-vars
+      // restrict development server's private groups to "blab-private"
+      .filter(([name, Source]) => PRODUCTION ? true : (Source.visibleToUser() || name==="blab-private"))
+      .map(([name, Source]) => [name, new Source()])
+  );
+
+  const datasetsPerGroup = new Map();
+  const narrativesPerGroup = new Map();
+
+  const updateKnownDatasetsAndNarratives = async () => {
+    // following loop is intentionally sequential to reduce load on server
+    for await (const [groupName, source] of groupSources) {
+      datasetsPerGroup.set(
+        groupName,
+        (await source.availableDatasets()).map((p) => `/groups/${groupName}/${p}`)
```

Comment:
These calls should be wrapped into `try/catch` block as stochastic network timeouts, running the server without appropriate credentials etc will will cause `availableDatasets` to throw. Similarly for `availableNarratives` a few lines later.

Reply:
I thought that we wouldn't need to due to https://github.com/nextstrain/nextstrain.org/pull/320?

Reply:
Is the difference here that the async requests for availableDatasets are actually happening outside of the scope of express handlers due to the details of the IIFE you explained above?

Reply:
> Is the difference here that the async requests for availableDatasets are actually happening outside of the scope of express handlers due to the details of the IIFE you explained above?

Yup!

Reply:
Just tested this to confirm, and since it's happening inside a promise (`updateKnownDatasetsAndNarratives` is an `async` function) which isn't being `await`ed, then any errors are `UnhandledPromiseRejectionWarning`; while these aren't fatal, we should be catching them.

If a `availableDatasets` method fails, we have some options for what to do:
1. If there were previous data set for this source, then keep the previous data. Data listing for certain groups will now be more out-of-date than others (but never more than ~24 hours, as [the Heroku server restarts every day](https://devcenter.heroku.com/articles/dynos#automatic-dyno-restarts)).
2. If there were previous data set for this source, then clear the previous data.

My preference is for (1) here. @eharkins?

Reply:
Fixed by @eharkins in 613ed72 👍